### PR TITLE
Squirrel it away

### LINKED ISSUE #2312: some Windows users not able to resolve the command line tool
xref https://github.com/desktop/desktop/issues/696#issuecomment-316569042

Known workaround: open a terminal and run this command:

```
%LocalAppData%\GitHubDesktop\app-0.7.0\GitHub Desktop.exe --squirrel-updated
```

Opening this up to track whether follow-up work is necessary.

### PR DESCRIPTION
**Pull request by joshaber**
Fixes #2312 

We were enforcing single instance even when we got a Squirrel event, which meant the updated instance would be quit before it had time to do anything.

Incidentally, this also caused a bug @niik had seen where the app would be focused after an update was downloaded.

### CODE REVIEW

File: app/src/lib/parse-app-url.ts [hunk 1]
```diff
@@ -42,7 +43,7 @@ export type URLActionType =
 export function parseAppURL(url: string): URLActionType {
   const parsedURL = URL.parse(url, true)
   const hostname = parsedURL.hostname
-  const unknown: IUnknownAction = { name: 'unknown' }
+  const unknown: IUnknownAction = { name: 'unknown', url }
```
Comment (joshaber) [inline app/src/lib/parse-app-url.ts]: Capturing the URL here makes our log more helpful when we can't parse the URL.

**Comment by joshaber** [comment]
🏷 🚀

File: app/src/main-process/main.ts [hunk 2]
```diff
@@ -43,100 +36,49 @@ const launchTime = now()
...
-    willQuit = true
+    handlingSquirrelEvent = true
     promise
-      .then(() => {
-        app.quit()
-      })
       .catch(e => {
         log.error(`Failed handling Squirrel event: ${arg}`, e)
       })
+      .then(() => {
```
Comment (shiftkey) [inline app/src/main-process/main.ts]: Does the order on `catch` and `then` matter here?

File: app/src/main-process/main.ts [hunk 3]
```diff
@@ -147,29 +89,37 @@ function handleAppURL(url: string) {
...
+    app.quit()
   }
-})
-
-if (isDuplicateInstance) {
-  willQuit = true
+}
 
-  app.quit()
+if (shellNeedsPatching(process)) {
```
Comment (shiftkey) [inline app/src/main-process/main.ts]: I'm curious (but not really surprised) about moving the patching after all the squirrel checks. Just playing it safe?

File: app/src/main-process/main.ts [hunk 4]
```diff
@@ -147,29 +89,37 @@ function handleAppURL(url: string) {
...
 }
 
-const isDuplicateInstance = app.makeSingleInstance((args, workingDirectory) => {
-  // Someone tried to run a second instance, we should focus our window.
-  if (mainWindow) {
-    if (mainWindow.isMinimized()) {
-      mainWindow.restore()
-    }
+let isDuplicateInstance = false
+// If we're handling a Squirrel event we don't wanna enforce single instance. We
```
Comment (shiftkey) [inline app/src/main-process/main.ts]: `wanna` -> `want to`

File: app/src/main-process/main.ts (hunk 2 above)
Comment (joshaber) [inline app/src/main-process/main.ts]: Yeah, with the `catch` before the `then` it ensures we always quit. The other way it wouldn't.

File: app/src/main-process/main.ts (hunk 3 above)
Comment (joshaber) [inline app/src/main-process/main.ts]: Playing it safe and also trying to group the launch logic together.

**Comment by joshaber** [comment]
🔰

**Comment by shiftkey** [comment]
:moon:

**Pull request merged**